决策树回归的自变量可以包括数值型和分类类型。

读取"D:\s-casa毕业论文-工程\R5R_TEST1_london\Test2_onmypoi_data\mergeResult\housing_hasId_accessibility.csv"，将property_type和borough列进行标签编码后，把编码的列加入features = ['bedrooms', 'bathrooms', 'size_square_meters', 
            'AccommodationAccess15m', 'AirportsAccess15m',
            'Culture and tourismAccess15m', 'Eating and drinkingAccess15m',
            'EducationAccess15m', 'EntertainmentAccess15m', 'HealthAccess15m',
            'Infrastructure and facilitiesAccess15m',
            'Manufacturing and productionAccess15m', 'OfficeAccess15m',
            'Public transportAccess15m', 'Road and railAccess15m',
            'ServiceAccess15m', 'ShoppingAccess15m', 'Water transportAccess15m']作为自变量，因变量是Price_per_square_meter列，做决策树回归。

In [2]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error, r2_score

# 读取CSV文件
file_path = r"D:\s-casa毕业论文-工程\R5R_TEST1_london\Test2_onmypoi_data\mergeResult\housing_hasId_accessibility.csv"
data = pd.read_csv(file_path)

# 提取自变量和因变量
features = ['bedrooms', 'bathrooms', 'size_square_meters', 
            'AccommodationAccess15m', 'AirportsAccess15m',
            'Culture and tourismAccess15m', 'Eating and drinkingAccess15m',
            'EducationAccess15m', 'EntertainmentAccess15m', 'HealthAccess15m',
            'Infrastructure and facilitiesAccess15m',
            'Manufacturing and productionAccess15m', 'OfficeAccess15m',
            'Public transportAccess15m', 'Road and railAccess15m',
            'ServiceAccess15m', 'ShoppingAccess15m', 'Water transportAccess15m']
target = 'Price_per_square_meter'

X = data[features]
y = data[target]

# 对property_type和borough列进行标签编码
label_encoder = LabelEncoder()
X['property_type_encoded'] = label_encoder.fit_transform(data['property_type'])
X['borough_encoded'] = label_encoder.fit_transform(data['borough'])
X

C:\Users\84398\AppData\Local\Temp\ipykernel_13716\1234828704.py:27: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X['property_type_encoded'] = label_encoder.fit_transform(data['property_type'])
C:\Users\84398\AppData\Local\Temp\ipykernel_13716\1234828704.py:28: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X['borough_encoded'] = label_encoder.fit_transform(data['borough'])


bedrooms  bathrooms  size_square_meters  AccommodationAccess15m  \
0            5          2                 212                       0   
1            3          1                 107                       0   
2            3          1                  82                       0   
3            3          1                  73                       1   
4            2          1                  85                       0   
...        ...        ...                 ...                     ...   
3969         2          2                 131                      10   
3970         3          2                 190                     168   
3971         1          1                  52                     123   
3972         3          2                 106                     130   
3973         1          1                  46                     108   

      AirportsAccess15m  Culture and tourismAccess15m  \
0                     0                             1   
1                     0                             8   
2                     0                             2   
3                     0                             1   
4                     0                             2   
...                 ...                           ...   
3969                  0                            56   
3970                  0                           188   
3971                  0                            46   
3972                  0                            84   
3973                  0                           125   

      Eating and drinkingAccess15m  EducationAccess15m  \
0                              141                   9   
1                               98                  10   
2                               23                   3   
3                               49                   7   
4                               10                   5   
...                            ...                 ...   
3969                           133                  16   
3970                          1043                  80   
3971                           262                  19   
3972                           500                  32   
3973                           695                  94   

      EntertainmentAccess15m  HealthAccess15m  \
0                         11               21   
1                         21               30   
2                          2                5   
3                          3               15   
4                          4               10   
...                      ...              ...   
3969                      19               47   
3970                     116              218   
3971                      20               49   
3972                      43               84   
3973                      90              189   

      Infrastructure and facilitiesAccess15m  \
0                                         26   
1                                         81   
2                                         19   
3                                         51   
4                                         31   
...                                      ...   
3969                                      54   
3970                                     283   
3971                                      81   
3972                                     140   
3973                                     254   

      Manufacturing and productionAccess15m  OfficeAccess15m  \
0                                         8                4   
1                                         8               10   
2                                         0                1   
3                                         4                2   
4                                         1                5   
...                                     ...              ...   
3969                                     13               26   
3970                                    139               83   
3971 

In [3]:



# 划分训练集和测试集
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 执行决策树回归
decision_tree_model = DecisionTreeRegressor(random_state=42)
decision_tree_model.fit(X_train, y_train)

# 预测
y_pred = decision_tree_model.predict(X_test)

# 评估模型
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("Mean Squared Error:", mse)
print("R-squared:", r2)


Mean Squared Error: 18918811.51743738
R-squared: 0.408005844606637


对于分类变量，由于在决策树回归中，字符串类型的特征无法直接用于建模，因为决策树回归算法处理的是数值型数据。

解决这个问题的方法是对字符串类型的特征进行预处理，将其转换为数值型数据，以便决策树回归算法能够处理它们。常见的方法是使用独热编码（One-Hot Encoding）或者标签编码（Label Encoding）。

如果特征只有两个类别，可以使用标签编码，将其转换为0和1。如果特征有多个类别，可以使用独热编码，将每个类别转换为一个独立的二进制特征。

In [12]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error, r2_score

# 读取CSV文件
file_path = r"D:\s-casa毕业论文-工程\R5R_TEST1_london\Test2_onmypoi_data\mergeResult\housing_hasId_accessibility.csv"
data = pd.read_csv(file_path)

# 提取自变量和因变量
features = ['bedrooms', 'bathrooms', 'size_square_meters', 
            'AccommodationAccess15m', 'AirportsAccess15m',
            'Culture and tourismAccess15m', 'Eating and drinkingAccess15m',
            'EducationAccess15m', 'EntertainmentAccess15m', 'HealthAccess15m',
            'Infrastructure and facilitiesAccess15m',
            'Manufacturing and productionAccess15m', 'OfficeAccess15m',
            'Public transportAccess15m', 'Road and railAccess15m',
            'ServiceAccess15m', 'ShoppingAccess15m', 'Water transportAccess15m',
            'property_type', 'borough']  # 包含了分类变量

target = 'Price_per_square_meter'

X = data[features]
y = data[target]

# 使用独热编码将分类变量转换为数值型特征
X = pd.get_dummies(X, columns=['property_type', 'borough'])
df_onehot = X

# 划分训练集和测试集
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 执行决策树回归
decision_tree_model = DecisionTreeRegressor(random_state=42)
decision_tree_model.fit(X_train, y_train)

# 预测
y_pred = decision_tree_model.predict(X_test)

# 评估模型
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("Mean Squared Error:", mse)
print("R-squared:", r2)


Mean Squared Error: 14785632.721441858
R-squared: 0.5373383710272267


In [8]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error, r2_score

# 定义决策树回归模型，并设置超参数的可能取值范围
tree_reg = DecisionTreeRegressor()
param_grid = {'max_depth': list(range(20,30)),
    'min_samples_split': list(range(1,8)),
    'min_samples_leaf': list(range(1, 6))}

# 使用GridSearchCV函数进行网格搜索
tree_reg_cv = GridSearchCV(tree_reg, param_grid, cv=5, scoring='neg_mean_squared_error')
tree_reg_cv.fit(X_train, y_train)


# 获取最佳模型和参数，并使用训练集和测试集对模型进行评估
best_tree_reg = tree_reg_cv.best_estimator_
print("最佳模型:", best_tree_reg)


# 在训练集上评估模型表现
y_train_pred = best_tree_reg.predict(X_train)
mse_train = mean_squared_error(y_train, y_train_pred)
rmse_train = np.sqrt(mse_train)
print("训练集MSE:", mse_train)
print("训练集RMSE:", rmse_train)

# 在训练集上使用模型进行预测，并计算R平方作为模型的评价指标。
y_pred = best_tree_reg.predict(X_train)
r2_train = r2_score(y_train, y_pred)
print("训练集R平方:", r2_train)

# 在测试集上评估模型表现
y_test_pred = best_tree_reg.predict(X_test)
mse = mean_squared_error(y_test, y_test_pred)
rmse = np.sqrt(mse)
print("测试集MSE:", mse)
print("测试集RMSE:", rmse)


# 在测试集上使用模型进行预测，并计算R平方作为模型的评价指标。
y_pred = best_tree_reg.predict(X_test)
r2_test = r2_score(y_test, y_pred)
print("测试集R平方:", r2_test)

最佳模型: DecisionTreeRegressor(max_depth=22, min_samples_leaf=5, min_samples_split=3)
训练集MSE: 5273176.8894674415
训练集RMSE: 2296.339889795812
训练集R平方: 0.8322946078372371
测试集MSE: 16238016.212036997
测试集RMSE: 4029.642194046141
测试集R平方: 0.49189140745714943


d:\programs\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:378: FitFailedWarning: 
250 fits failed out of a total of 1750.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
250 fits failed with the following error:
Traceback (most recent call last):
  File "d:\programs\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "d:\programs\anaconda3\lib\site-packages\sklearn\tree\_classes.py", line 1247, in fit
    super().fit(
  File "d:\programs\anaconda3\lib\site-packages\sklearn\tree\_classes.py", line 177, in fit
    self._validate_params()
  File "d:\programs\anaconda3\lib\site-packages\sklearn\base.py", line 600, in _va

最佳模型: DecisionTreeRegressor(max_depth=22, min_samples_leaf=5, min_samples_split=3)

训练集MSE: 5273176.8894674415
训练集RMSE: 2296.339889795812
训练集R平方: 0.8322946078372371

测试集MSE: 16238016.212036997
测试集RMSE: 4029.642194046141
测试集R平方: 0.49189140745714943

在相关性分析中，我们通常将相关系数的绝对值大小来判断变量之间的相关程度。相关系数的取值范围在-1到1之间，表示了变量之间的线性关系强度和方向。

当相关系数为1时，表示两个变量完全正相关，意味着它们之间存在着完全的线性正相关关系。这种情况下，一个变量的增加将导致另一个变量的增加，并且变量之间的散点图呈现完全的上升趋势。

当相关系数为-1时，表示两个变量完全负相关，意味着它们之间存在着完全的线性负相关关系。这种情况下，一个变量的增加将导致另一个变量的减少，并且变量之间的散点图呈现完全的下降趋势。

当相关系数接近0时，表示两个变量之间没有线性相关关系，它们之间的散点图没有明显的趋势，或者说它们之间的线性关系非常弱。

根据您提供的相关系数结果，可以观察到以下情况：

'Price_per_square_meter' 与 'borough_Kensington and Chelsea' 之间的相关系数约为 0.54，表示这两个变量之间存在一定程度的正相关关系。

'Price_per_square_meter' 与 'ShoppingAccess15m' 之间的相关系数约为 0.45，也表示这两个变量之间存在一定程度的正相关关系。

'Price_per_square_meter' 与其他变量的相关系数较小（绝对值接近于0），意味着它们之间的线性相关关系较弱或者基本没有线性相关。

In [18]:
# 假设 df_onehot 是经过独热编码后的数据集
# 假设 data 是原始数据集，包含 'Price_per_square_meter' 列

# 提取目标变量 'Price_per_square_meter' 列的数据
target_column = 'Price_per_square_meter'
target_data = data[target_column]

# 将目标变量数据添加为一列到 df_onehot 数据集中
df_onehot[target_column] = target_data

# 现在 df_onehot 中包含了 'Price_per_square_meter' 列
df_onehot

bedrooms  bathrooms  size_square_meters  AccommodationAccess15m  \
0            5          2                 212                       0   
1            3          1                 107                       0   
2            3          1                  82                       0   
3            3          1                  73                       1   
4            2          1                  85                       0   
...        ...        ...                 ...                     ...   
3969         2          2                 131                      10   
3970         3          2                 190                     168   
3971         1          1                  52                     123   
3972         3          2                 106                     130   
3973         1          1                  46                     108   

      AirportsAccess15m  Culture and tourismAccess15m  \
0                     0                             1   
1                     0                             8   
2                     0                             2   
3                     0                             1   
4                     0                             2   
...                 ...                           ...   
3969                  0                            56   
3970                  0                           188   
3971                  0                            46   
3972                  0                            84   
3973                  0                           125   

      Eating and drinkingAccess15m  EducationAccess15m  \
0                              141                   9   
1                               98                  10   
2                               23                   3   
3                               49                   7   
4                               10                   5   
...                            ...                 ...   
3969                           133                  16   
3970                          1043                  80   
3971                           262                  19   
3972                           500                  32   
3973                           695                  94   

      EntertainmentAccess15m  HealthAccess15m  ...  borough_Newham  \
0                         11               21  ...           False   
1                         21               30  ...           False   
2                          2                5  ...           False   
3                          3               15  ...           False   
4                          4               10  ...           False   
...                      ...              ...  ...             ...   
3969                      19               47  ...           False   
3970                     116              218  ...           False   
3971                      20               49  ...           False   
3972                      43               84  ...           False   
3973                      90              189  ...           False   

      borough_Redbridge  borough_Richmond upon Thames  borough_Southwark  \
0                 False                         False              False   
1                 False                         False              False   
2                 False                         False              False   
3                 False                         False              False   
4                 False                         False              False   
...                 ...                           ...                ...   
3969              False                         False              False   
3970              False                         False              False   
3971              False                         False              False   
3972              False                         False              False   
3973              False                         False              Fal

In [23]:


# 假设 df_onehot 是经过独热编码后的数据集
# 计算相关系数矩阵
correlation_matrix = df_onehot.corr()

# 相关系数矩阵是一个对称矩阵，我们可以通过查看每对变量的相关系数来进行分析
# 例如，查看某一列变量与其他变量的相关系数
target_column = 'Price_per_square_meter'
correlations_with_target = correlation_matrix[target_column].sort_values(ascending=False)

pd.set_option('display.max_rows', None) 

print("与目标变量相关性排名：")
print(correlations_with_target)

pd.set_option('display.max_rows', 10) 

与目标变量相关性排名：
Price_per_square_meter                        1.000000
borough_Kensington and Chelsea                0.544359
ShoppingAccess15m                             0.450547
Culture and tourismAccess15m                  0.449064
AccommodationAccess15m                        0.430802
HealthAccess15m                               0.388329
OfficeAccess15m                               0.361803
Eating and drinkingAccess15m                  0.348315
ServiceAccess15m                              0.338443
EntertainmentAccess15m                        0.320798
Infrastructure and facilitiesAccess15m        0.302787
EducationAccess15m                            0.301605
Manufacturing and productionAccess15m         0.291607
Public transportAccess15m                     0.291021
borough_Westminster                           0.287409
bathrooms                                     0.263210
property_type_Apartment                       0.183442
Road and railAccess15m                        0.17740